### PR TITLE
page search and filers + view hijack page

### LINKED ISSUE #38: Implement Hijack page
@vkotronis can you add a screenshot of the hijack page? 

This is the code from old frontend
https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/frontend/webapp/render/templates/main/hijack.htm

### LINKED ISSUE #39: Add view more on bgp and hijack table
@vkotronis can you add a screenshot from the expanded bgp update and the more link on hijacks table?

### LINKED ISSUE #40: Add filtering and search on BGP/Hijack page

### PR DESCRIPTION
**Pull request by CuriouzK0d3r**
resolves #38 
resolves #39 
resolves #40

### CODE REVIEW

File: apps/artemis-web/components/hijack-table/hijack-table.tsx [hunk 1]
```diff
@@ -1,5 +1,32 @@
...
+  className: 'my-custom-text-filter', // custom classname on input
+  defaultValue: '', // default filtering value
+  comparator: Comparator.EQ, // default is Comparator.LIKE
+  caseSensitive: true, // default is false, and true will only work when comparator is LIKE
+  style: {}, // your custom styles on input
+  delay: 1000, // how long will trigger filtering after user typing, default is 500 ms
+  id: 'id', // assign a unique value for htmlFor attribute, it's useful when you have same dataField across multiple table in one page
+});
+
+const selectRPKI = {
```
Comment (slowr) [inline apps/artemis-web/components/hijack-table/hijack-table.tsx:22]: why is this needed if it's 1-1? is there no way to do a toString?

File: apps/artemis-web/components/bgp-table/bgp-table.tsx [hunk 2]
```diff
@@ -1,5 +1,127 @@
...
+  className: 'my-custom-text-filter', // custom classname on input
+  defaultValue: '', // default filtering value
+  comparator: Comparator.EQ, // default is Comparator.LIKE
+  caseSensitive: true, // default is false, and true will only work when comparator is LIKE
+  style: {}, // your custom styles on input
+  delay: 1000, // how long will trigger filtering after user typing, default is 500 ms
+  id: 'id', // assign a unique value for htmlFor attribute, it's useful when you have same dataField across multiple table in one page
+});
+
+const selectOptions = {
```
Comment (slowr) [inline apps/artemis-web/components/bgp-table/bgp-table.tsx:21]: why is this needed if it's 1-1? is there no way to do a toString?

File: apps/artemis-web/components/hijack-table/hijack-table.tsx [hunk 3]
```diff
@@ -183,23 +218,38 @@ const HijackTableComponent = (props) => {
...
       ASes: row.num_asns_inf,
-      ack: row.seen,
-      more: row.comment,
+      ack:
+        row.resolved || row.under_mitigation ? (
+          <img src="./handled.png" />
+        ) : (
+          <img src="./unhadled.png" />
+        ),
+      more: <a href={'/hijack?key=' + row.key}>View</a>,
```
Comment (slowr) [inline apps/artemis-web/components/hijack-table/hijack-table.tsx:237]: If we want to visit this page with SSR, would it make sense to just pass the components as props?

Saying that I want both ones to be supported, so you either pass it from the parameter in the URL or from the component prop.

File: apps/artemis-web/components/ongoing-hijack-table/ongoing-hijack-table.tsx [hunk 4]
```diff
@@ -1,34 +1,105 @@
...
+  className: 'my-custom-text-filter', // custom classname on input
+  defaultValue: '', // default filtering value
+  comparator: Comparator.EQ, // default is Comparator.LIKE
+  caseSensitive: true, // default is false, and true will only work when comparator is LIKE
+  style: {}, // your custom styles on input
+  delay: 1000, // how long will trigger filtering after user typing, default is 500 ms
+  id: 'id', // assign a unique value for htmlFor attribute, it's useful when you have same dataField across multiple table in one page
+});
+
+const selectOptions = {
```
Comment (slowr) [inline apps/artemis-web/components/ongoing-hijack-table/ongoing-hijack-table.tsx:22]: why is this needed if it's 1-1? is there no way to do a toString?

File: apps/artemis-web/components/status-table/status-table.tsx [hunk 5]
```diff
@@ -27,7 +52,14 @@ class StatusTable extends Component<StatsType, {}> {
             STATS_DATA.view_processes.map((process, i) => {
               return (
                 <tr key={i}>
-                  <td>{process.name}</td>
+                  <td
+                    data-toggle="tooltip"
+                    data-placement="top"
+                    title={tooltips[process.name]}
```
Comment (slowr) [inline apps/artemis-web/components/status-table/status-table.tsx:58]: if this doesn't exist then we don't have an error right?

File: apps/artemis-web/pages/bgpupdates.tsx [hunk 6]
```diff
@@ -50,11 +51,26 @@ const BGPUpdates = (props) => {
   const bgp = BGP_DATA ? BGP_DATA.view_bgpupdates : [];
   const filteredDate = new Date();
   filteredDate.setHours(filteredDate.getHours() - filter);
-
-  const filteredBgp =
+  console.log(bgp);
```
Comment (slowr) [inline apps/artemis-web/pages/bgpupdates.tsx:54]: remove

File: apps/artemis-web/pages/hijack.tsx [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    else if (row.under_mitigation) return 'Under Mitigation';
+    else if (row.outdated) return 'Outdated';
+    else return '';
+  };
+
+  const classes = useStyles();
+  const router = useRouter();
+  const key: any = router.query.key;
+  const user = props.user;
+
```
Comment (slowr) [inline apps/artemis-web/pages/hijack.tsx:67]: missing a key null check

File: apps/artemis-web/pages/hijacks.tsx [hunk 8]
```diff
@@ -28,34 +28,74 @@ const useStyles = makeStyles((theme) => ({
...
+    else if (row.resolved) return 'Resolved';
+    else if (row.ignored) return 'Ignored';
+    else if (row.active) return 'Active';
+    else if (row.dormant) return 'Dormant';
+    else if (row.under_mitigation) return 'Under Mitigation';
+    else if (row.outdated) return 'Outdated';
+    else return '';
+  };
+
+  const setStatus = (id) => {
```
Comment (slowr) [inline apps/artemis-web/pages/hijacks.tsx:72]: why pass an ID and not have the name directly? The statuses list does not make sense.

File: apps/artemis-web/pages/hijack.tsx [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    Ongoing: 'danger',
+    Dormant: 'secondary',
+    Resolved: 'success',
+    Ignored: 'warning',
+    'Under Mitigation': 'primary',
+    Withdrawn: 'info',
+    Outdated: 'dark',
+  };
+
+  const findStatus = (row) => {
```
Comment (slowr) [inline apps/artemis-web/pages/hijack.tsx:52]: There can be multiple statuses like withdrawn+outdated, etc
These are the possible states:
https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/other/db/data/tables.sql#L83

**Review by slowr — CHANGES_REQUESTED** [review]
Very good work so far. Please check the comments for further changes!

File: apps/artemis-web/components/status-table/status-table.tsx (hunk 5 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/components/status-table/status-table.tsx:58]: If its not existent, I now assign an empty string to the tooltip.

File: apps/artemis-web/pages/hijacks.tsx [hunk 10]
```diff
@@ -112,64 +138,247 @@ const HijacksPage = (props) => {
...
+                            onChange={(event) => setKey(event.target.value)}
+                            className="form-control"
+                            placeholder="Type hijack key..."
+                            id="hijack-key-input"
+                          />
+                        </div>
+                        <div className="col-sm-1">
+                          <button
+                            onClick={() =>
+                              window.location.replace('/hijack?key=' + key)
```
Comment (slowr) [inline apps/artemis-web/pages/hijacks.tsx:216]: We want SSR on this

**Review by slowr — CHANGES_REQUESTED** [review]

File: apps/artemis-web/components/404-hoc/404-hoc.tsx [hunk 11]
```diff
@@ -19,7 +19,7 @@ const NotFoundHOC = (WrappedComponent, ACL = []) => {
             <meta name="robots" content="noindex" />
           </Head>
           <DefaultErrorPage
-            statusCode={404}
+            statusCode={401}
```
Comment (slowr) [inline apps/artemis-web/components/404-hoc/404-hoc.tsx:22]: 404 hoc on the title but it returns 401? :)

File: apps/artemis-web/pages/hijack.tsx [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+    if (row.dormant) statuses.push('Dormant');
+    if (row.under_mitigation) statuses.push('Under Mitigation');
+    if (row.outdated) statuses.push('Outdated');
+
+    return statuses;
+  };
+
+  const classes = useStyles();
+  const router = useRouter();
+  const key: string = props.key ?? router.query.key ?? '';
```
Comment (slowr) [inline apps/artemis-web/pages/hijack.tsx:61]: since our last discussion we said we won't pass it from props

**Review by slowr — CHANGES_REQUESTED** [review]

**Comment by slowr** [comment]
Some of the things I found while playing on the bgpupdates page:

Got this while playing around with pages/time periods:
image

Hovering doesn't work it seems always, mostly on the AS paths; sometimes just saying true
image
image

Show X entries doesn't actually show X entries.

A lot of console errors! React errors, trying to find ASundefined, etc
image
image

Show X entries is without margin from the test and Download Table button is too close to the table
image

Table starts from page 0

File: apps/artemis-web/pages/api/userlist.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import {
+  NextApiRequestExtended,
+  NextApiResponseExtended,
+} from '../../definitions';
+import auth from '../../middleware/auth';
+
+const handler = nc()
+  .use(auth)
+  .get(async (req: NextApiRequestExtended, res: NextApiResponseExtended) => {
+    if (!req.user || req.user.role !== 'admin') {
```
Comment (slowr) [inline apps/artemis-web/pages/api/userlist.ts:11]: hmm, is it possible to add this check in the .use chain like a middleware?
something like:
```
.use(auth)
.use(authorization(['admin']))
.get(...)
```

File: apps/artemis-web/pages/api/usermanagement.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
+import nc from 'next-connect';
+import {
+  NextApiRequestExtended,
+  NextApiResponseExtended,
+} from '../../definitions';
+import auth from '../../middleware/auth';
+
+const handler = nc()
+  .use(auth)
+  .post(async (req: NextApiRequestExtended, res: NextApiResponseExtended) => {
```
Comment (slowr) [inline apps/artemis-web/pages/api/usermanagement.ts:10]: same comment as the other

**Review by slowr — CHANGES_REQUESTED** [review]
wow! great looking PR. Look at my last comment and we should be able to merge!

File: apps/artemis-web/pages/api/userlist.ts (hunk 13 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/pages/api/userlist.ts:11]: Nice idea! Simple and elegant.

File: apps/artemis-web/utils/fetch-data.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    'https://stat.ripe.net/data/abuse-contact-finder/data.json?resource=AS' +
+      ASN
+  ).then((response) => response.json());
+}
+
+function getData(ASN) {
+  return Promise.all([getName(ASN), getCountry(ASN), getAbuse(ASN)]);
+}
+
+export function fetchASNData(ASN) {
```
[the remark this answers is not in the document]
Comment (slowr) [reply apps/artemis-web/utils/fetch-data.ts:25]: Now I see that it fetches the ASN data on the page refresh. Imagine having 1M BGP updates, it is going to destroy ripe by doing all the queries. Is there no way to do it only on mouse-over like we had it on legacy frontend?
https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/frontend/webapp/render/static/js/custom/ripestat.js#L19

[the remark this answers is not in the document]
Comment (slowr) [reply apps/artemis-web/utils/fetch-data.ts:25]: Lastly, we the current implementation does it work if we are getting new updates? will it query on every new bgp update?

[the remark this answers is not in the document]
Comment (slowr) [reply apps/artemis-web/utils/fetch-data.ts:25]: And we want the ASN info also on the AS PATH field :)

**Review by slowr — CHANGES_REQUESTED** [review]
These seems to be the last comments that I have.

**Review by slowr — APPROVED** [review]
